components/activity_and_insights/insights/widget_hoc/widget_hoc.tsx

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import React, {ComponentType, useCallback} from 'react';
+import {useIntl} from 'react-intl';
+
+import {CardSize} from '../insights';
+import InsightsCard from '../card/card';
+
+import {InsightsScopes, InsightsCardTitles} from 'utils/constants';
+
+export interface WidgetHocProps {
+    size: CardSize;
+    widgetType: 'TOP_CHANNELS' | 'TOP_REACTIONS';
+    filterType: string;
+    class: string;
+}
+
+function widgetHoc<T>(WrappedComponent: ComponentType<T>) {
+    const Component = (props: T & WidgetHocProps) => {
+        const {formatMessage} = useIntl();
+
+        const title = useCallback(() => {
+            if (props.filterType === InsightsScopes.MY) {
+                return formatMessage(InsightsCardTitles[props.widgetType].myTitle);
+            }
+            return formatMessage(InsightsCardTitles[props.widgetType].teamTitle);
+        }, [props.filterType, props.widgetType]);
+
+        const subTitle = useCallback(() => {
+            if (props.filterType === InsightsScopes.MY) {
+                return formatMessage(InsightsCardTitles[props.widgetType].mySubTitle);
+            }
+            return formatMessage(InsightsCardTitles[props.widgetType].teamSubTitle);
+        }, [props.filterType, props.widgetType]);
+
+        return (
+            <InsightsCard
+                class={props.class}
+                title={title()}
+                subTitle={subTitle()}
+                size={props.size}
+            >
+                <WrappedComponent
+                    {...(props as unknown as T)}
+                />
+            </InsightsCard>
+
+        );
+    };
```

Comment:
For the title and subtitle, could we use a callback that is passed into the `WrappedComponent` and called with the title and subtitle? 

```
getTitle = (title: string, subTitle: string) => {
     ... use state to update title and subtitle of InsightsCardTitles component
}
...
<WrappedComponent
    getTitle={this.getTitle}, ...otherProps
>
```
Not 100% if this would work, or align with how you're designing these components, but it could help with removing the need for `widgetType`.

Reply:
Good idea but we need to keep widget type for the modal in a later PR. Also, keeping the title/subtitle functions out of the parent component is better due to a separation of concerns and to not bloat the parent component